hr_expense_cancel/models/hr_expense.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2019 Tecnativa - Ernesto Tejeda
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import models
+
+
+class HrExpenseSheet(models.Model):
+    _inherit = "hr.expense.sheet"
+
+    def action_cancel(self):
+        for sheet in self:
+            account_move = sheet.account_move_ids
+            sheet.account_move_ids = False
+            payments = sheet.payment_ids.filtered(lambda line: line.state != "cancel")
+            # case : cancel invoice from hr_expense
+            self._remove_reconcile_hr_invoice(account_move)
+            # If the sheet is paid then remove payments
+            if sheet.state in ("done", "approve"):
+                if sheet.expense_line_ids[:1].payment_mode == "own_account":
+                    self._remove_move_reconcile(payments, account_move)
+                    self._cancel_payments(payments)
+                else:
+                    # In this case, during the cancellation the journal entry
+                    # will be deleted
+                    self._cancel_payments(payments)
+            # Deleting the Journal entry if in the previous steps
+            # (if the expense sheet is paid and payment_mode == 'own_account')
+            # it has not been deleted
+            if account_move.exists():
+                move_to_cancel = account_move.filtered(
+                    lambda move: move.state != "draft"
+                )
+                if move_to_cancel:
+                    move_to_cancel.button_cancel()
+                account_move.with_context(force_delete=True).unlink()
+            sheet.state = "submit"
+
+    def _remove_reconcile_hr_invoice(self, account_move):
+        """Cancel invoice made by hr_expense_invoice module automatically"""
+        reconcile = account_move.mapped("line_ids.full_reconcile_id")
+        aml = self.env["account.move.line"].search(
+            [("full_reconcile_id", "in", reconcile.ids)]
+        )
+        exp_move_line = aml.filtered(
+            lambda line: line.move_id.id not in account_move.ids
+        )
```

Comment:
Why run a `search()` only to filter out records right after? All this could be implemented via `search()` only:
```suggestion
        exp_move_line = self.env["account.move.line"].search(
            [
                ("full_reconcile_id", "in", account_move.line_ids.full_reconcile_id.ids),
                ("move_id", "not in", account_move.ids),
            ]
        )
```